pkg/products/monitoringspec/reconciler.go

```diff
@@ -0,0 +1,388 @@
+		return nil
+	})
+	if err != nil {
+		return err
+	}
+	r.Logger.Infof("operation result of creating servicemonitor %v was %v", sm.Name, opRes)
+
+	//Get the service monitor - that was created/updated
+	sermon := &v1.ServiceMonitor{}
+	err = serverClient.Get(ctx, k8sclient.ObjectKey{Name: sm.Name, Namespace: r.Config.GetNamespace()}, sermon)
+	if err != nil {
+		return err
+	}
+	//Create role binding for each of the namespace label selectors
+	for _, namespace := range sermon.Spec.NamespaceSelector.MatchNames {
+		err := r.reconcileRoleBindings(ctx, serverClient, namespace)
+		if err != nil {
+			return err
+		}
+	}
+	return err
+}
+
+func (r *Reconciler) reconcileRoleBindings(ctx context.Context,
+	serverClient k8sclient.Client, namespace string) (err error) {
+
+	roleBinding := &rbac.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      roleBindingName,
+			Namespace: namespace,
+		},
+	}
+	opRes, err := controllerutil.CreateOrUpdate(ctx, serverClient, roleBinding, func() error {
+		roleBinding.Subjects = []rbac.Subject{
+			{
+				Kind:      rbac.ServiceAccountKind,
+				Name:      clusterMonitoringPrometheusServiceAccount,
+				Namespace: clusterMonitoringNamespace,
+			},
+		}
+		roleBinding.RoleRef = rbac.RoleRef{
+			APIGroup: roleRefAPIGroup,
+			Kind:     bundle.ClusterRoleKind,
+			Name:     roleRefName,
+		}
+		return nil
+	})
+	r.Logger.Infof("operation result of creating rolebinding: %v was %v", roleBindingName, opRes)
+	return err
+}
+
+func (r *Reconciler) removeServicemonitor(ctx context.Context,
+	serverClient k8sclient.Client, namespace, name string) (err error) {
+	//Get the service monitor
+	sm := &v1.ServiceMonitor{}
+	err = serverClient.Get(ctx, k8sclient.ObjectKey{Name: name, Namespace: namespace}, sm)
+	if err != nil && k8serr.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	//Delete the servicemonitor
+	err = serverClient.Delete(ctx, sm)
+	if err != nil && k8serr.IsNotFound(err) {
+		return nil
+	}
+	return err
+}
+
+func (r *Reconciler) removeRoleBinding(ctx context.Context,
+	serverClient k8sclient.Client, namespace, name string) (err error) {
+
+	// Check if the namespace has service monitors
+	// if so donot delete the rolebinding
+	listOpts := []k8sclient.ListOption{
+		k8sclient.InNamespace(namespace),
+	}
+	serviceMonitors := &v1.ServiceMonitorList{}
+	err = serverClient.List(ctx, serviceMonitors, listOpts...)
+	if err != nil {
+		return err
+	}
+	if len(serviceMonitors.Items) > 0 {
+		return nil
+	}
+
+	//Get the rolebinding
+	rb := &rbac.RoleBinding{}
+	err = serverClient.Get(ctx, k8sclient.ObjectKey{Name: name, Namespace: namespace}, rb)
+	if err != nil && k8serr.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	//Delete the rolebinding
+	err = serverClient.Delete(ctx, rb)
+	if err != nil && k8serr.IsNotFound(err) {
+		return nil
+	}
+	return err
+}
+
+func (r *Reconciler) getServiceMonitors(ctx context.Context,
```

Comment:
There are a couple of places in this change where listing ServiceMonitors is done without using this function. I think either this function should be split or those other places should learn to use it. What do you think?

Reply:
Agreed. Re factored the getServicemonitors to take a list options as argument. Fixed.